ngclient: Fix intermediate metadata loading with rollback checks

jku opened the pull request:
The rollback checks themselves work, but they create a situation
where Updater does not realize that it needs to download e.g. a new
snapshot because the local snapshot is valid as _intermediate_ snapshot
(that can be used for rollback protection but nothing else), but is not
valid as final snapshot.

The PR changes very little but documentation is improved in several places:
* defines `_check_final_timestamp()` for consistency with `_check_final_snapshot()`
* `update_timestamp()` and `update_snapshot()` will still load "intermediate" metadata as before, but will now raise in the end of function if the metadata is not valid final metadata
* this is what Updater is expecting so there are no changes in Updater

The reason this bug was introduced (by me) was lacking tests. The tests are still missing: I found the issue while working on the tests in #1444. It might make sense to merge this commit when the tests are merged but I wanted to have separate review for this as it's clearly an implementation bug.

Commit:
ngclient: Fix rollback checks

The rollback checks themselves work, but they create a situation
where Updater does not realize that it needs to download e.g. a new
snapshot because the local snapshot is valid as _intermediate_ snapshot
(that can be used for rollback protection but nothing else), but is not
valid as final snapshot.

Raise in the end of update_snapshot and update_timestamp if the files
are not valid final metadata: this way the intermediate metadata does
get loaded but Updater also knows it is not the final metadata.

This modifies the existing tests but does not yet test the situation
described in the first paragraph.

Fixes #1563

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
The other way of solving this would have been to return something from the `update_*()` methods to indicate whether metadata is valid final metadata or not. I chose to not do that because this way keeps Updater implementation simple:  a single try-except around the local metadata loading is enough to decide if loading remote metadata is needed

joshuagl reviewed (approved):
Tidy little change, LGTM.

sechkova reviewed (approved):
I approve the set of code changes and I agree that this approach is better than using the update_* methods return values. 

I think we should in addition make some effort to explain the new terminology that we are adopting: intermediate timestamp, snapshot etc. are not explicitly mentioned in the specification and are not exactly the same as intermediate root versions. The same metadata is referred to as "trusted" in 5.3.11 of the spec,  the current client uses the notion of previous/current while seen from Updater this is the loaded local metadata.
Not to be misunderstood, I'm ok with intermediate but we need a definition somewhere.

Commit:
TrustedMetadataSet: Improve module docstring

Explain what "intermediate" metadata is and how it affects the loading
process.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

joshuagl commented on tuf/ngclient/_internal/trusted_metadata_set.py:
```diff
@@ -13,11 +13,18 @@
+ * Metadata must be loaded in order:
+   root -> timestamp -> snapshot -> targets -> (delegated targets).
+ * Metadata can be loaded even if it is expired (or in the snapshot case if the
+   meta info does not match): this is called "intermediate metadata".
```
Nit: it might be worth mentioning rollback protection explicitly here?

jku commented:
Closing since this merged in #1569
